mailpoet/assets/js/src/automation/integrations/mailpoet/steps/send_email/edit/edit_newsletter.tsx

```diff
@@ -5,9 +5,31 @@ import { Button } from '../../../components/button';
 import { storeName } from '../../../../../editor/store';
 import { MailPoet } from '../../../../../../mailpoet';
 
+const emailPreviewLinkCache = {};
```

Comment:
Using a local memory cache like this feels a little bit off to me in a react/redux context even though I think it works perfectly fine as far as I can tell. I know you don't want to make duplicate API calls every time an email step is opened for the n+1th time though. 

Might it make sense to store the retrieved preview URLs in redux instead to persist the data between mounts?

I'm not sure it fully makes sense to me to to fire off an API call every single time an email step is opened for the first time just to get the preview URL. It's not exactly an expensive operation or anything, and it doesn't block anything else, but what percentage of the time will users actually want to use that link?

If someone is switching quickly between tabs you could also end up in a situation where the email step component unmounts before the ajax request finishes, and then the `setPreviewLink` callback won't work and might emit a warning, though probably not with react in production mode.

I might consider having the preview button work more similarly to the create button, where clicking the button fires off the ajax request to get the preview URL (or retrieves it from the cache/redux/wherever) and then opens a new tab, just to prevent unnecessary API calls.

Anyway, this does work fine for now, just options to consider. @JanJakes please let us know if you can think of any actual issues we might encounter from using a cache like this, since you're much more familiar with the implications of this kind of stuff from a react perspective.

Reply:
I do not like the solution as well. For the feature itself, I thought creating its own store or extending an existing one is a bit too much though (at least for now). But open to suggestions :)

Reply:
@johnolek @websupporter Usually, using a module global to store data for React components is risky because when you have multiple instances of the same component, they would override each others data. But in this case, it may actually work to it's advantage — multiple instances of the same component could read from the same cache.

But there's another issue — the `retrievePreviewLink` is called in the component body, not wrapped in an `useEffect` which will very likely produce JS errors in some cases:
```ts
  if (emailId && !previewLink) {
    void retrievePreviewLink(emailId, setPreviewLink);
  }
```

Plus, even if the above is fixed, multiple usages of the same component are likely to trigger the request in parallel.

Other ideas:
1. Fetch preview URL on click, then use it, as @johnolek said.
2. Add a new non-public URL for "email preview by ID" that would redirect to the preview URL.
3. Replicate the URL generation logic in JS (we may forget to sync it with PHP, if ever changed).
4. Save `preview_url` in args (but this can break old emails if we ever changed the logic somehow).

What do you think? I'm not sure what I prefer, option 2. seems kind of reasonable, I guess.

Reply:
Thanks for your feedback. I opted for the first option in [857ab8f](https://github.com/mailpoet/mailpoet/pull/4455/commits/857ab8fbf080889fa03115a4e80f21de0784124e)